File: ingest/internal/tracing/tracing.go
// Package tracing wires OpenTelemetry into the ingestion tier.
//
// The point of tracing here is not the ingestion tier on its own — that is
// one process and its logs are legible. The point is that a trace started
// when a venue frame arrives survives the hop through Kafka and is picked up
// again by the serving tier, so "why was this quote stale" can be answered
// across three processes and two languages in one view instead of by
// correlating timestamps by hand.
//
// Propagation is W3C Trace Context carried in Kafka record headers. That
// choice matters: the alternative — embedding trace ids in the Avro payload —
// would put an observability concern into the data schema, and every schema
// evolution would then have to reason about it.
package tracing

import (
	"context"
	"fmt"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
	"go.opentelemetry.io/otel/trace/noop"
)

// ServiceName is the resource name this tier reports as.
const ServiceName = "tapeline-ingestd"

// Config controls the exporter.
type Config struct {
	// Endpoint is the OTLP gRPC collector address. Empty disables tracing
	// entirely and installs a no-op tracer.
	Endpoint string
	// SampleRatio in [0,1]. Market data is high-volume: tracing every event
	// would cost more than the pipeline it observes, and a head sample of a
	// few percent answers the same questions.
	SampleRatio float64
	Version     string
}

// Shutdown flushes buffered spans. Always call it — an exporter that is not
// shut down drops whatever is still in its queue, which is reliably the
// spans from the incident you care about.
type Shutdown func(context.Context) error

// Init installs the global tracer provider and propagator.
func Init(ctx context.Context, cfg Config) (Shutdown, error) {
	// Tracing off is a supported configuration, not a degraded one. The local
	// dry-run mode has no collector and should not need one.
	if cfg.Endpoint == "" {
		otel.SetTracerProvider(noop.NewTracerProvider())
		return func(context.Context) error { return nil }, nil
	}

	exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(cfg.Endpoint),
		otlptracegrpc.WithInsecure(),
		otlptracegrpc.WithTimeout(5*time.Second),
	)
	if err != nil {
		return nil, fmt.Errorf("otlp exporter: %w", err)
	}

	res, err := resource.Merge(
		resource.Default(),
		resource.NewWithAttributes(
			semconv.SchemaURL,
			semconv.ServiceName(ServiceName),
			semconv.ServiceVersion(cfg.Version),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("otel resource: %w", err)
	}

	ratio := cfg.SampleRatio
	if ratio <= 0 || ratio > 1 {
		ratio = 0.01
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter,
			sdktrace.WithMaxQueueSize(4096),
			sdktrace.WithBatchTimeout(5*time.Second),
		),
		sdktrace.WithResource(res),
		// ParentBased so a sampled trace stays sampled end to end. Sampling
		// independently at each hop produces traces with holes in them, which
		// are worse than no trace at all.
		sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.TraceIDRatioBased(ratio))),
	)

	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown, nil
}

// Tracer returns the ingestion tier's tracer.
func Tracer() trace.Tracer { return otel.Tracer(ServiceName) }

// KafkaHeaderCarrier adapts a string map to the OTel propagator interface so
// trace context can ride in Kafka record headers.
type KafkaHeaderCarrier map[string]string

func (c KafkaHeaderCarrier) Get(key string) string { return c[key] }
func (c KafkaHeaderCarrier) Set(key, value string) { c[key] = value }
func (c KafkaHeaderCarrier) Keys() []string {
	keys := make([]string, 0, len(c))
	for k := range c {
		keys = append(keys, k)
	}
	return keys
}

// InjectKafka serializes the span context in ctx into headers a Kafka record
// can carry. Returns nil when nothing is being traced, so an untraced
// publish adds no header bytes at all.
func InjectKafka(ctx context.Context) map[string]string {
	if !trace.SpanContextFromContext(ctx).IsValid() {
		return nil
	}
	carrier := KafkaHeaderCarrier{}
	otel.GetTextMapPropagator().Inject(ctx, carrier)
	if len(carrier) == 0 {
		return nil
	}
	return carrier
}

// ExtractKafka rebuilds a span context from Kafka headers. The serving tier
// does the equivalent on the Java side.
func ExtractKafka(ctx context.Context, headers map[string]string) context.Context {
	if len(headers) == 0 {
		return ctx
	}
	return otel.GetTextMapPropagator().Extract(ctx, KafkaHeaderCarrier(headers))
}

// VenueAttributes are the attributes every ingestion span carries.
func VenueAttributes(venue, symbol, kind string) []attribute.KeyValue {
	return []attribute.KeyValue{
		attribute.String("tapeline.venue", venue),
		attribute.String("tapeline.symbol", symbol),
		attribute.String("tapeline.kind", kind),
	}
}

File: serving/src/main/java/io/tapeline/serving/tracing/Tracing.java
package io.tapeline.serving.tracing;

import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Context;
import io.opentelemetry.context.propagation.ContextPropagators;
import io.opentelemetry.context.propagation.TextMapGetter;
import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
import io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
import io.opentelemetry.sdk.trace.samplers.Sampler;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.header.Header;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * The consumer half of the distributed trace.
 *
 * <p>The Go ingestion tier writes W3C Trace Context into Kafka record headers
 * rather than into the Avro payload, precisely so that a consumer in a
 * different language can pick it up without the schema knowing anything about
 * observability. This class is the other end of that contract: it extracts the
 * `traceparent` header off a consumed record and continues the same trace.
 *
 * <p>The result is one trace spanning a venue WebSocket read in Go, a Kafka
 * hop, and a Redis write in Java — which is the only way to answer "why was
 * this quote stale" without correlating timestamps across three log streams by
 * hand.
 *
 * <p>Tracing is off unless {@code TAPELINE_OTLP_ENDPOINT} is set. That is a
 * supported configuration, not a degraded one: the unit tests and the local
 * dry-run have no collector and must still work.
 */
@Configuration
public class Tracing {

    public static final String SERVICE_NAME = "tapeline-serving";

    /** Reads Kafka record headers for the propagator. */
    private static final TextMapGetter<ConsumerRecord<?, ?>> KAFKA_GETTER =
            new TextMapGetter<>() {
                @Override
                public Iterable<String> keys(ConsumerRecord<?, ?> record) {
                    List<String> keys = new ArrayList<>();
                    for (Header h : record.headers()) {
                        keys.add(h.key());
                    }
                    return keys;
                }

                @Override
                public String get(ConsumerRecord<?, ?> record, String key) {
                    if (record == null) {
                        return null;
                    }
                    Header header = record.headers().lastHeader(key);
                    return header == null
                            ? null
                            : new String(header.value(), StandardCharsets.UTF_8);
                }
            };

    @Bean
    public OpenTelemetry openTelemetry(
            @Value("${tapeline.otlp.endpoint:}") String endpoint,
            @Value("${tapeline.otlp.sample-ratio:0.05}") double sampleRatio) {

        if (endpoint == null || endpoint.isBlank()) {
            // No collector configured. A no-op propagates nothing and costs
            // nothing, and every call site below stays unconditional.
            return OpenTelemetry.noop();
        }

        Resource resource = Resource.getDefault().merge(
                Resource.create(Attributes.of(
                        AttributeKey.stringKey("service.name"), SERVICE_NAME)));

        SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
                .setResource(resource)
                // ParentBased so a trace sampled by the Go tier stays sampled
                // here. Sampling independently at each hop yields traces with
                // holes, which are worse than no trace at all.
                .setSampler(Sampler.parentBased(Sampler.traceIdRatioBased(sampleRatio)))
                .addSpanProcessor(BatchSpanProcessor.builder(
                                OtlpGrpcSpanExporter.builder()
                                        .setEndpoint("http://" + endpoint)
                                        .setTimeout(Duration.ofSeconds(5))
                                        .build())
                        .setScheduleDelay(Duration.ofSeconds(5))
                        .build())
                .build();

        return OpenTelemetrySdk.builder()
                .setTracerProvider(tracerProvider)
                .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
                .buildAndRegisterGlobal();
    }

    @Bean
    public Tracer tracer(OpenTelemetry openTelemetry) {
        return openTelemetry.getTracer(SERVICE_NAME);
    }

    /**
     * Continues the trace a Kafka record carries, or starts a fresh root when
     * the record has no trace headers — which is what a record written before
     * tracing existed looks like.
     */
    public static Context extract(OpenTelemetry otel, ConsumerRecord<?, ?> record) {
        return otel.getPropagators()
                .getTextMapPropagator()
                .extract(Context.current(), record, KAFKA_GETTER);
    }
}
